resources/Microsoft.NET.SDK.ToolInstaller/Microsoft.NET.SDK.ToolInstaller.psm1

```diff
@@ -0,0 +1,404 @@
+    [bool] $PreRelease = $false
+
+    [DscProperty()]
+    [string] $ToolPath
+
+    [DscProperty()]
+    [bool] $Exist = $true
+
+    static [hashtable] $InstalledPackages
+
+    NETSDKToolInstaller()
+    {
+        [NETSDKToolInstaller]::GetInstalledPackages()
+    }
+
+    NETSDKToolInstaller([string] $PackageId, [string] $Version, [string[]] $Commands, [bool] $PreRelease, [string] $ToolPath)
+    {
+        $this.PackageId = $PackageId
+        $this.Version = $Version
+        $this.Commands = $Commands
+        $this.PreRelease = $PreRelease
+        $this.ToolPath = $ToolPath
+    }
+
+    [NETSDKToolInstaller] Get()
+    {
+        $properties = $this.ToHashTable()
+
+        $installed = [NETSDKToolInstaller]::Export($properties)
+
+        $currentState = $installed | Where-Object { $_.PackageId -eq $this.PackageId }
+        if ($null -ne $currentState)
+        {
+            # update the export list
+            $currentState::InstalledPackages[$this.PackageId] = $currentState
+            return $currentState
+        }
+        
+        return [NETSDKToolInstaller]@{
+            PackageId = $this.PackageId
+            Version   = $this.Version
+            Commands  = $this.Commands
+            Exist     = $false
+        }
+    }
+
+    Set()
+    {
+        # TODO: validate for upgrade/update scenarios
+        if ($this.Test())
+        {
+            return
+        }
+
+        if ($this.Exist)
+        {
+            $this.Install($false)
+        }
+        else
+        {
+            $this.Uninstall($false)
+        }
+    }
+
+    [bool] Test()
+    {
+        $currentState = $this.Get()
+        if ($currentState.Exist -ne $this.Exist)
+        {
+            return $false
+        }
+
+        if ($null -ne $this.Version -and $this.Version -ne $currentState.Version)
+        {
+            return $false
+        }
+
+        return $true
+    }
+
+    static [NETSDKToolInstaller[]] Export()
+    {
+        return [NETSDKToolInstaller]::Export(@{})
+    }
+
+    static [NETSDKToolInstaller[]] Export([hashtable] $filterProperties)
+    {
+        $command = "tool list --global"
+        
+        $toolsDir = $filterProperties.ContainsKey("ToolPath")
+        if ($toolsDir)
+        {
+            $command = "tool list --tool-path $($filterProperties.ToolPath)"
+        }
+ 
+        $packageList = Invoke-DotNet -Command $command
+
+        $inputObject = Convert-DotNetToolOutput -Output $packageList
+
+        $results = [List[NETSDKToolInstaller]]::new()
+
+        foreach ($package in $inputObject)
+        {
+            # determine if the package is a pre-release package
+            $pre = $false
+            $preReleasePackage = $package.Version -Split "-"
+            if ($preReleasePackage.Count -gt 1)
+            {
+                # set the pre-release flag to true to build the object
+                $pre = $true
+            }
+
+            # default directory
+            $defaultDir = " "
```

Comment:
nit: Is this the same as `$installDirectory`?  That seems a little bit more clear to me but just my opinion.